awesome/src/objects/tag.rs

```diff
@@ -34,8 +36,87 @@ impl Default for TagState {
+    }
+
+    pub fn get_clients(&self) -> rlua::Result<Table<'lua>> {
+        self.0.table()?.get("__clients")
+    }
+
+    pub fn set_clients(&mut self, clients: rlua::Table) -> rlua::Result<Value> {
+        // TODO: this is a really inefficient solution ( O(n*m) for the search )
+        //   Since it does not generally treat big arrays, this may be acceptable,
+        //   However a faster algorithm would not hurt if someone finds one
+
+        let prev_clients = self.get_clients()?;
+
+        // Indexes refering to prev_clients
+        let mut not_common_clients: Vec<u32> = (1..(prev_clients.len()? + 1) as u32)
+                                                    .collect();
+        let mut num_common_clients = 0;
+            
+        for client in clients.clone().sequence_values::<AnyUserData>() {
+            let mut idx = -1;
+            let client = Client::cast(client?.clone().into())?;
+            let self_ref = &*client.state()? as *const _;
+
+            for pair in prev_clients.clone().pairs::<Integer, AnyUserData>() {
+                let (i, c) = pair?;
+                if &*c.borrow::<ClientState>()? as *const _ == self_ref {
+                    idx = i;
+                    num_common_clients += 1;
+                    not_common_clients.remove((i - num_common_clients) as usize);
+                    break;
+                }
+            }
+            if idx != -1 {
+                // TODO: emit signal
+            }
+        }
+        for idx in not_common_clients.iter() {
+            // prev_clients.get::<Integer, ClientState>(idx)?
+            // TODO: emit signal and garbage if not referenced anymore
+        }
+        self.0.table()?.set("__clients", clients)?;
+        Ok(Value::Nil)
+    }
+
+    pub fn client_index(&self, client: &Client) -> rlua::Result<Integer> {
+        let clients = self.get_clients()?;
+        let self_ref = &*client.state()? as *const _;
+        for pair in clients.pairs::<Integer, AnyUserData<'lua>>() {
+            let (i, c) = pair?;
+            if &*c.borrow::<ClientState>()? as *const _ == self_ref {
+                return Ok(i);
+            }
+        }
+        Err(rlua::Error::RuntimeError("Client not found".to_string()))
+    }
+
+    pub fn tag_client(&mut self, obj: AnyUserData<'lua>) -> rlua::Result<Value> {
+        let client = Client::cast(obj.clone().into())?;
+        if let Ok(_) = self.client_index(&client) {
+            return Ok(Value::Nil);
+        }
+        let clients = self.get_clients()?;
+        clients.set(clients.len()? + 1, obj)?;
+        self.set_clients(clients)?;
+        Ok(Value::Nil)
+    }
+
+    pub fn untag_client(&mut self, obj: AnyUserData<'lua>) -> rlua::Result<Value> {
+        let client = Client::cast(obj.clone().into())?;
+        let index = self.client_index(&client)?;
+        let clients = self.get_clients()?;
+
+        for i in index..clients.len()? {
+            clients.set(i, clients.get::<_, Value>(i + 1)?)?;
```

Comment:
Don't do this. Store `__clients` as a `Vec<AnyUserdata>`. [Vec has an auto impl ToLua provided by the rlua crate](https://docs.rs/rlua/0.15.3/rlua/prelude/trait.FromLua.html#impl-FromLua%3C%27lua%3E). Then you can simply cast it to a `Vec<ClientState>` and not have to do this weirdness.

Reply:
If you need guidance on this, look at how `screen.rs` stores screen handles in its `SCREEN_HANDLE`. Just grep `SCREEN_HANDLE` in `screen.rs`.